File: cloud/lib/vaccine-display-name.ts
/**
 * Adds a manufacturer prefix to COVID vaccine names for display (Will
 * 2026-09-24, verbatim: "Vaccine names in assist app: Add Pfizer and
 * Moderna to the names for Spikevax, mNexspike, and Comirnaty. Start
 * with mfg name then go to the drug name itself, only on covid
 * vaccines."). Pure/no I/O, display-only — the underlying stored name
 * (matched by lib/vaccine-product-catalog.ts and
 * app/api/ordering/recommendation/route.ts, aliased by the screener,
 * and read verbatim by the desktop app) must never change, so callers
 * apply this ONLY where a vaccine name is rendered to a human, never
 * to a value used for matching, sorting keys, API payloads, or stored
 * data.
 *
 * Case-insensitive on the match; the season/age suffix already on the
 * name (e.g. "Comirnaty 2026-27 12+") is preserved verbatim after the
 * prefix. Idempotent: a name that already starts with "Pfizer" or
 * "Moderna" (any case) is returned unchanged rather than double-prefixed.
 * Every non-COVID name passes through untouched.
 *
 * ROUND 2 (macro codes follow-up, Will 2026-09-24 verbatim: "vaccine
 * macro codes: follow up to brand covnetion. Make it look liek this:
 * Pfizer 12+ (Comirnaty 2026-27)"): exports covidVaccineMaker below,
 * the maker-lookup half of vaccineDisplayName's logic, so
 * lib/macro-codes.ts's covidMacroLabel can reuse the exact same
 * COVID-product detection/maker mapping instead of duplicating it.
 * vaccineDisplayName itself is unchanged (still used verbatim
 * everywhere else in the app).
 *
 * ROUND 3 (reviewer finding on covidMacroLabel, 2026-09-24): also
 * exports stripCovidMakerPrefix below — the ALREADY_PREFIXED-handling
 * half of covidVaccineMaker, on its own, so a caller that needs the
 * bare drug name (not just the maker) can strip an already-present
 * "Pfizer "/"Moderna " prefix before it does its own parsing, instead
 * of re-detecting that prefix itself.
 */

const COVID_PREFIXES: { test: RegExp; prefix: string }[] = [
  { test: /^comirnaty/i, prefix: "Pfizer" },
  { test: /^spikevax/i, prefix: "Moderna" },
  { test: /^mnexspike/i, prefix: "Moderna" },
];

const ALREADY_PREFIXED = /^(pfizer|moderna)\b\s*/i;

/** Strips a leading "Pfizer "/"Moderna " prefix (any case, and any
 * amount of whitespace after it) from `name`, if present — "Pfizer
 * Comirnaty 2026-27" -> "Comirnaty 2026-27". A name with no such
 * prefix passes through unchanged. */
export function stripCovidMakerPrefix(name: string): string {
  return name.replace(ALREADY_PREFIXED, "");
}

/** Returns "Pfizer"/"Moderna" for a Comirnaty/Spikevax/mNEXSPIKE name
 * (case-insensitive, and tolerant of a name that's already
 * maker-prefixed — see stripCovidMakerPrefix above), or null for any
 * non-COVID name. The shared maker lookup behind vaccineDisplayName
 * below and lib/macro-codes.ts's covidMacroLabel. */
export function covidVaccineMaker(name: string): string | null {
  const nameToTest = stripCovidMakerPrefix(name);
  for (const { test, prefix } of COVID_PREFIXES) {
    if (test.test(nameToTest)) return prefix;
  }
  return null;
}

export function vaccineDisplayName(name: string): string {
  if (ALREADY_PREFIXED.test(name)) return name;

  const maker = covidVaccineMaker(name);
  return maker ? `${maker} ${name}` : name;
}

File: desktop/VaccineAssist.Desktop/Services/VaccineDisplayName.cs
using System.Text.RegularExpressions;

namespace VaccineAssist.Desktop.Services;

/// <summary>
/// Mirrors cloud/lib/vaccine-display-name.ts's vaccineDisplayName exactly
/// (Will, V-T55, 2026-09-25 verbatim: "It should show the same names
/// everywhere that it is applicable.") — adds a manufacturer prefix to
/// COVID vaccine names for DISPLAY ONLY. The desktop app calls
/// GET /api/vaccines and GET /api/physician-rules directly (see
/// Services/VaccineApiService.cs) and gets back the raw stored name; the
/// cloud side's own displayName addition to those JSON payloads is a
/// separate, parallel change this does NOT depend on, so the rule is
/// re-implemented here rather than read off the wire.
///
/// Case-insensitive on the match; any season/age suffix already on the
/// name (e.g. "Comirnaty 2026-27 12+") is preserved verbatim after the
/// prefix. Idempotent: a name that already starts with "Pfizer"/"Moderna"
/// (any case) is returned unchanged rather than double-prefixed. Every
/// non-COVID name passes through untouched. Null/empty-safe: never throws,
/// returns "" for a null/empty input.
///
/// Callers MUST apply this ONLY where a vaccine name is rendered to a
/// human — never to a value used for matching, sorting keys, API request
/// payloads, PioneerRx automation input, or anything persisted/looked up
/// by name. See VaccineAssist.Desktop.Common.VaccineDisplayNameConverter
/// for the XAML-binding equivalent of this same rule.
/// </summary>
public static class VaccineDisplayName
{
    private static readonly Regex AlreadyPrefixed =
        new(@"^(pfizer|moderna)\b\s*", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    private static readonly (Regex Test, string Prefix)[] CovidPrefixes =
    {
        (new Regex(@"^comirnaty", RegexOptions.IgnoreCase | RegexOptions.Compiled), "Pfizer"),
        (new Regex(@"^spikevax", RegexOptions.IgnoreCase | RegexOptions.Compiled), "Moderna"),
        (new Regex(@"^mnexspike", RegexOptions.IgnoreCase | RegexOptions.Compiled), "Moderna"),
    };

    /// <summary>Returns <paramref name="name"/> with "Pfizer "/"Moderna "
    /// prefixed when it names a COVID product (Comirnaty/Spikevax/
    /// mNEXSPIKE), unchanged otherwise. Never null.</summary>
    public static string For(string? name)
    {
        if (string.IsNullOrEmpty(name)) return name ?? "";
        if (AlreadyPrefixed.IsMatch(name)) return name;

        foreach (var (test, prefix) in CovidPrefixes)
        {
            if (test.IsMatch(name)) return $"{prefix} {name}";
        }

        return name;
    }
}
